feat: updated desktop modeler plugin documentation (#687)
camunda-cloud/camunda-cloud-documentation: docs/components/modeler/desktop-modeler/plugins/plugins.md

The Camunda Modeler searches for available plug-in entry points via the `resources/plugins/*/index.js` glob. That being said, each plug-in must reside in it's own folder which is a direct child of the `plugins` directory. If you download and extract plugins from GitHub the extracted directory contains the actual plugin, so make sure to copy the plugin, not its parent directory.

<br>

## Overview of your possibilities as a plugin developer

There are many way you as a developer are able to extend our modeling tools, and since this ecosystem is a little large here's an overview of the ways you can customize our software:

| Plugin type               | Functionality                                                                                                        | Example                                                                                                                         |
| ------------------------- | -------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- |

> | Menu Entries              | Add new entries to the menu bar - useful to interact with your plugins, link to external pages, add settings, ect... | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/camunda/camunda-modeler-plugins/tree/master/menu-plugin-example)                     |

Reviewer: This applies to all of the following lines

| Custom Styles             | Change the look and feel of Camunda Modeler and your extensions                                                      | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/camunda/camunda-modeler-plugins/tree/master/style-plugin-example)                    |

> | React Components          | Embed custom React components into specific anchor points in our application                                         | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/pinussilvestrus/camunda-modeler-autosave-plugin)                                     |

Reviewer: Sometimes you talk about `our application` sometimes about `Camunda Modeler`. Please make it consistent. I recommend:
```suggestion
| React Components          | Embed custom React components into specific anchor points of the Camunda Modeler                                       | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/pinussilvestrus/camunda-modeler-autosave-plugin)                                     |
```

This also applies to the other table entries

| bpmn-js Modules           | Extend our BPMN editor by creating your own custom modules                                                           | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/camunda/camunda-modeler-plugins/tree/master/bpmn-js-plugin-example)                  |
| bpmn-js Moddle Extensions | Enable the use of custom BPMN semantics                                                                              | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/camunda/camunda-modeler-plugins/tree/master/bpmn-js-plugin-moddle-extension-example) |
| dmn-js Modules            | Extend our DMN editor by creating your own custom modules                                                            | &nbsp;&nbsp;&nbsp;[**↗**](https://github.com/camunda/camunda-modeler-plugins/tree/master/dmn-js-plugin-example)                   |

> | dmn-js Moddle Extensions  | Enable the use of custom DMN semantics                                                                               | &nbsp;&nbsp;&nbsp;n/a |

Reviewer: ```suggestion
| dmn-js Moddle Extensions  | Extend the DMN language model by injecting your own custom [dmn-moddle](https://github.com/bpmn-io/dmn-moddle) modules                                                                             | &nbsp;&nbsp;&nbsp;n/a |
```

  name: 'My Awesome Plugin', // the name of your plugin
  style: './style.css', // changing the appearance of the modeler
  menu: './menu.js', // adding menu entries to the modeler

>   script: './script.js' // extend the modeler, and its BPMN and DMN components

Reviewer: ```suggestion
  script: './script.js' // extending the modeler, and its BPMN and DMN components
```

```javascript
module.exports = {
  menu: './menu-entries'
};
```

> > The code within the menu entries is special because it executes on [the main process](https://www.electronjs.org/docs/latest/tutorial/process-model) of Electron. This comes with the advantage of allowing you to use [Node.js](https://nodejs.org/en/) modules, and the disadvantage of not being able to debug your code in Chromium. For more information regarding main process debugging, please refer to the [official Electron documentation](https://www.electronjs.org/docs/latest/tutorial/debugging-main-process).

Author: @MaxTru Added this entry, thoughts?

Reviewer: I like. 

I would propose a slight rephrasing (not a disadvantage, but a challenge to consider ;-) ). Something like:

```suggestion
> The code within the menu entries executes on [the main process](https://www.electronjs.org/docs/latest/tutorial/process-model) of Electron. This comes with the advantage of allowing you to use [Node.js](https://nodejs.org/en/) modules, but you need to consider, that you can not debug the respective code in Chromium. For more information regarding main process debugging, please refer to the [official Electron documentation](https://www.electronjs.org/docs/latest/tutorial/debugging-main-process).
```

:::

For more information on how the modeler's menu works, take a look at its [implementation](https://github.com/camunda/camunda-modeler/blob/master/app/lib/menu/menu-builder.js).

### Extend the modeler and its BPMN and DMN components

You can extend the modeling tools for [BPMN](https://github.com/bpmn-io/bpmn-js) and [DMN](https://github.com/bpmn-io/dmn-js) with your own modules, as well as embedding React.js components into certain sections of Camunda Modeler.

Since the client of the modeler uses [Chromium](https://www.chromium.org/Home), you can't use Node.js modules to extend the modeling tools. You need to bundle your plugin first. The easiest way to get started with client-side plugins is through [this example project](https://github.com/camunda/camunda-modeler-plugin-example).

> > In this example, we are building a bpmn-js plugin, but this basic structure applies to all extensions besides menu entries and style. The modules themselves will be different however, so refer to our [examples](https://github.com/camunda/camunda-modeler-plugins) for more information on how to build different kinds.

Reviewer 2: "...how to build different kinds of modules?"

registerBpmnJSPlugin(plugin);
```

You may want to create a plugin which specifically targets Camunda 7 or Camunda 8. To do this use the appropriate variations of the registration helper function for your plugin type.

```javascript

> // Such variations are currently only available for BPMN but will soon become available for DMN as well
> registerPlatformBpmnJSPlugin(plugin)

Reviewer: ```suggestion
registerPlatformBpmnJSPlugin(plugin); // Register plugin for Camunda 7 BPMN diagrams only
registerCloudBpmnJSPlugin(plugin); // Register plugin for Camunda 8 BPMN diagrams only
registerBpmnJSPlugin(plugin); // Register plugin for Camunda 7 and 8 BPMN diagrams
```
